presto-main/src/main/java/com/facebook/presto/operator/MergeOperator.java

```diff
@@ -0,0 +1,268 @@
+    {
+        private final int operatorId;
+        private final PlanNodeId sourceId;
+        private final ExchangeClientSupplier exchangeClientSupplier;
+        private final PagesSerdeFactory serdeFactory;
+        private final List<Type> types;
+        private final List<Integer> outputChannels;
+        private final List<Type> outputTypes;
+        private final List<Integer> sortChannels;
+        private final List<SortOrder> sortOrder;
+        private final PageWithPositionComparatorFactory pageWithPositionComparatorFactory;
+        private LoadingCache<URI, ExchangeClient> exchangeClientCache;
+        private boolean closed;
+
+        public MergeOperatorFactory(
+                int operatorId,
+                PlanNodeId sourceId,
+                ExchangeClientSupplier exchangeClientSupplier,
+                PagesSerdeFactory serdeFactory,
+                PageWithPositionComparatorFactory pageWithPositionComparatorFactory,
+                List<Type> types,
+                List<Integer> outputChannels,
+                List<Integer> sortChannels,
+                List<SortOrder> sortOrder)
+        {
+            this.operatorId = operatorId;
+            this.sourceId = requireNonNull(sourceId, "sourceId is null");
+            this.exchangeClientSupplier = requireNonNull(exchangeClientSupplier, "exchangeClientSupplier is null");
+            this.serdeFactory = requireNonNull(serdeFactory, "serdeFactory is null");
+            this.types = requireNonNull(types, "types is null");
+            this.outputChannels = requireNonNull(outputChannels, "outputChannels is null");
+            this.outputTypes = outputTypes(types, outputChannels);
+            this.sortChannels = requireNonNull(sortChannels, "sortChannels is null");
+            this.sortOrder = requireNonNull(sortOrder, "sortOrder is null");
+            this.pageWithPositionComparatorFactory = requireNonNull(pageWithPositionComparatorFactory, "pageWithPositionComparatorFactory is null");
+        }
+
+        private static List<Type> outputTypes(List<? extends Type> sourceTypes, List<Integer> outputChannels)
+        {
+            return mappedCopy(outputChannels, sourceTypes::get);
+        }
+
+        @Override
+        public PlanNodeId getSourceId()
+        {
+            return sourceId;
+        }
+
+        @Override
+        public SourceOperator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+            OperatorContext operatorContext = driverContext.addOperatorContext(operatorId, sourceId, MergeOperator.class.getSimpleName());
+
+            if (exchangeClientCache == null) {
+                exchangeClientCache = CacheBuilder.<URI, ExchangeClient>newBuilder()
+                        .build(CacheLoader.from(location -> {
+                            ExchangeClient exchangeClient = exchangeClientSupplier.get(driverContext.getPipelineContext().localSystemMemoryContext());
```

Comment:
I may be missing something here, but why do we need to cache exchange clients? Is it really that expensive to create a new exchange client?

Reply:
We don't want to have multiple clients for the same location though. Make sure that's not the case.

Reply:
> We don't want to have multiple clients for the same location though

There is just one `MergeOperator` instance. I think this should be guaranteed by query scheduler that the same split is not passed twice. I looked at your original commit @arhimondr and there wasn't any cache there, so I think I must have introduced it to cope with pipeline system memory context which appears to be unnecessary.